### PR TITLE
Implement async_copy!

### PR DESCRIPTION
**Pull request by vchuravy**
bors r+

### CODE REVIEW

File: src/backends/cuda.jl [hunk 1]
```diff
@@ -47,39 +47,60 @@ end
...
-# do a CPU wait, and therefore block anyone from submitting more work.
-# We maybe could do a spinning wait on the GPU and atomic flag to signal from the CPU,
-# but which stream would we target?
-wait(::CUDA, ev::CPUEvent,  progress=nothing) = wait(CPU(), ev, progress)
+    return CudaEvent(event)
+end
 
-function __enqueue_wait(ev::CudaEvent, stream::CuStream)
-    CUDAdrv.wait(ev.event, stream)
+function pin!(::CUDA, a)
```
Comment (leios) [inline src/backends/cuda.jl]: Do the users actually need to see the `pin!(...)` function? I feel like we should just do it for the GPU-> GPU copy and not show anyone. This function is naturally inefficient on the CPU, so having the user allocate unused memory seems a bit weird.

Comment (lcw) [reply src/backends/cuda.jl]: Yes, I think users need to see `pin!`.  Some users may want to manage their own data transfers between the host and device.  Also, I don't understand why this method is inefficient on the CPU.

File: src/backends/cpu.jl [hunk 2]
```diff
@@ -15,10 +15,33 @@ function wait(::CPU, ev::CPUEvent, progress=nothing)
...
         end
     end
 end
+function __waitall(::CPU, dependencies, progress)
+    if dependencies isa Event
+        dependencies = (dependencies,)
+    end
+    if dependencies !== nothing
+        dependencies = collect(dependencies)
+        cpudeps  = filter(d->d isa CPUEvent,    dependencies)
```
Comment (lcw) [inline src/backends/cpu.jl]:
```suggestion
         cpudeps = filter(d->d isa CPUEvent && d.task isa Core.Task, dependencies)
```

File: src/backends/cuda.jl (hunk 1 above)
Comment (vchuravy) [reply src/backends/cuda.jl]: The issue is that `pin!` is costly, but we could cache the pointer...

Comment (leios) [reply src/backends/cuda.jl]: I just noticed there is no `pin!(::CPU, a)` function in this PR right now, so I guess if it's a GPU-specific function, that's fine. I didn't want users to allocate a new array to pin in the sake of language agnostic code.

File: src/KernelAbstractions.jl [hunk 3]
```diff
@@ -3,6 +3,8 @@ module KernelAbstractions
 export @kernel
 export @Const, @localmem, @private, @uniform, @synchronize, @index, groupsize
 export Device, GPU, CPU, CUDA, Event
+export async_copy!, pin!
```
Comment (lcw) [inline src/KernelAbstractions.jl]:
```suggestion
export async_copy!
```

File: test/async_copy.jl [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    event = async_copy!(backend, A, a, dependencies=event)
+    wait(event)
+
+    @test isapprox(a, Array(A))
+    @test isapprox(a, Array(B))
+end
+
+M = 1024
+
+if has_cuda_gpu()
```
Comment (leios) [inline test/async_copy.jl]: Maybe add testsets around these.

Also, I feel there should be a test for all possible copies: CPU-> CPU, GPU->CPU, GPU->GPU (with pin), CPU->GPU.

Comment (vchuravy) [reply test/async_copy.jl]: Interestingly the tests to hit all those code-paths already!

**Pull request merged**